presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,162 @@
+import sun.misc.SignalHandler;
+
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.cli.ConsolePrinter.REAL_TERMINAL;
+import static com.google.common.base.Strings.emptyToNull;
+import static com.google.common.base.Throwables.propagateIfPossible;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class QueryPreprocessor
+{
+    public static final String ENV_PREPROCESSOR = "PRESTO_PREPROCESSOR";
+    public static final String ENV_PREPROCESSOR_TIMEOUT = "PRESTO_PREPROCESSOR_TIMEOUT";
+    public static final String ENV_PRESTO_CATALOG = "PRESTO_CATALOG";
+    public static final String ENV_PRESTO_SCHEMA = "PRESTO_SCHEMA";
+    private static final Duration DEFAULT_PREPROCESSOR_TIMEOUT = new Duration(10, SECONDS);
+
+    private static final Signal SIGINT = new Signal("INT");
+    private static final String PREPROCESSING_QUERY_MESSAGE = "Preprocessing query";
+
+    private QueryPreprocessor() {}
+
+    public static String preprocessQuery(Optional<String> catalog, Optional<String> schema, String query)
+            throws QueryPreprocessorException
+    {
+        Duration timeout = DEFAULT_PREPROCESSOR_TIMEOUT;
+        String timeoutEnvironment = System.getenv(ENV_PREPROCESSOR_TIMEOUT);
+        if (emptyToNull(timeoutEnvironment) != null) {
+            timeout = Duration.valueOf(timeoutEnvironment.trim());
+        }
+
+        String preprocessorCommand = System.getenv(ENV_PREPROCESSOR);
+        if (emptyToNull(preprocessorCommand) == null) {
+            return query;
+        }
+        return preprocessQuery(catalog, schema, query, ImmutableList.of("/bin/sh", "-c", preprocessorCommand.trim()), timeout);
+    }
+
+    public static String preprocessQuery(Optional<String> catalog, Optional<String> schema, String query, List<String> preprocessorCommand, Duration timeout)
+            throws QueryPreprocessorException
+    {
+        Thread clientThread = Thread.currentThread();
+        SignalHandler oldHandler = Signal.handle(SIGINT, signal -> clientThread.interrupt());
+        try {
+            if (REAL_TERMINAL) {
+                System.out.print(PREPROCESSING_QUERY_MESSAGE);
+                System.out.flush();
+            }
+            return preprocessQueryInternal(catalog, schema, query, preprocessorCommand, timeout);
+        }
+        finally {
+            if (REAL_TERMINAL) {
+                System.out.print("\r" + Strings.repeat(" ", PREPROCESSING_QUERY_MESSAGE.length()) + "\r");
+                System.out.flush();
+            }
+            Signal.handle(SIGINT, oldHandler);
+            Thread.interrupted(); // clear interrupt status
```

Comment:
I copied this pattern from the other cli code that uses the signal handler, but I think it makes sense. We may have set the interrupted flag on the thread, and we want the thread to be in the default state.